das/coordinator.go

```diff
@@ -0,0 +1,139 @@
+type samplingCoordinator struct {
+	concurrencyLimit int
+	sampleFn         sampleFn
+	store            *store
+
+	state coordinatorState
+
+	resultCh chan result // fan-in sampling results from worker to coordinator
+	updMaxCh chan uint64 // signals to update max known header height
+	waitCh   chan func() // blocks coordinator for external access to state
+
+	workersWg       sync.WaitGroup
+	workersDone     done
+	coordinatorDone done
+}
+
+type result struct {
+	job
+	failed []uint64
+	err    error
+}
+
+func newSamplingCoordinator(
+	concurrency int,
+	sample sampleFn,
+	store *store) *samplingCoordinator {
+	return &samplingCoordinator{
+		concurrencyLimit: concurrency,
+		sampleFn:         sample,
+		store:            store,
+		resultCh:         make(chan result),
+		updMaxCh:         make(chan uint64),
+		waitCh:           make(chan func()),
+		workersDone:      newDone("workers"),
+		coordinatorDone:  newDone("sampling coordinator"),
+	}
+}
+
+func (sm *samplingCoordinator) run(ctx context.Context) {
+	for {
+		for !sm.concurrencyLimitReached() {
+			next, found := sm.state.nextJob()
+			if !found {
+				break
+			}
+			sm.runWorker(ctx, next)
+		}
+
+		select {
+		case max := <-sm.updMaxCh:
+			sm.state.updateMaxKnown(max)
+		case res := <-sm.resultCh:
+			sm.state.handleResult(res)
+		case wait := <-sm.waitCh:
+			wait()
+		case <-ctx.Done():
+			sm.coordinatorDone.Done()
+			sm.workersWg.Wait()
+			sm.workersDone.Done()
+			return
+		}
+	}
+}
+
+// runWorker runs job in separate worker go-routine
+func (sm *samplingCoordinator) runWorker(ctx context.Context, j job) {
+	w := worker{}
+	sm.state.putInProgress(j.id, w.getState)
+
+	// launch worker go-routine
+	sm.workersWg.Add(1)
+	go func() {
+		defer sm.workersWg.Done()
+		w.run(ctx, j, sm.sampleFn, sm.resultCh)
+	}()
+}
+
+// finished stores last sampling state to internal storage and waits for all routines to finish
+func (sm *samplingCoordinator) finished(ctx context.Context) error {
+	// wait for coordinator to exit and store state
+	if err := sm.coordinatorDone.wait(ctx); err != nil {
+		return err
+	}
+	if err := sm.store.store(ctx, newCheckpoint(sm.state.stats())); err != nil {
+		log.Errorw("storing checkpoint to disk", "Err", err)
+	}
+
+	// wait for all worker routines to finish
+	if err := sm.workersDone.wait(ctx); err != nil {
+		return err
+	}
+	if err := sm.store.store(ctx, newCheckpoint(sm.state.stats())); err != nil {
+		log.Errorw("storing checkpoint to disk", "Err", err)
+	}
+	return nil
+}
+
+// listen notifies coordinator about new header height has been found
+func (sm *samplingCoordinator) listen(ctx context.Context, height uint64) {
+	select {
+	case sm.updMaxCh <- height:
+	case <-ctx.Done():
+	}
+}
+
+// listens for new headers to keep maxHeader up-to-date
+func (sm *samplingCoordinator) getStats(ctx context.Context) (SamplingStats, error) {
+	var wg sync.WaitGroup
+	wg.Add(1)
+	defer wg.Done()
+
+	select {
+	case sm.waitCh <- wg.Wait:
```

Comment:
Just for understanding: So every time stats are collected (for example, for the background store), you stop spawning new workers until all jobs are completed, then you get the stats, and start spinning up workers again?

Reply:
It's a fancy way to lock the coordinator's routine and access its state. 
> Fancy because I haven't seen it before

Reply:
Fanciness or cleverness should be avoided if possible if it can be done in a simpler way - but that is a different topic

I still don't know the answer to my question, can you hold my hand? Does locking the coordinator's routine then mean that no new workers are spawned until all jobs are finished, so stats can be collected, and then the routine can continue?

Reply:
My main goal was to make it as simple as possible tbh. 
Regarding first question - no. Workers are spawn until concurrencyLimit is reached. Then coordinator awaits for results from any worker before will spin out another one, if there are jobs available. In the meantime while workers are busy doing their jobs, corrdinator is not blocked from updating `maxKnown` or exporting stats to any external request.

This piece of code is responsible for preventing race condition on collecting stats from corrdinator. It blocks coordinator with waitGroup untill stats has been read and unblocks after by releasing waitGroup. This operation is also running on demand, while coordinator is not busy starting new workers or updating state.

Reply:
> Fanciness or cleverness should be avoided if possible if it can be done in a simpler way - but that is a different topic

Obviously, but any other method we can think of(mutexes or req/resp) does not apply in this case that well. It is not just _fancy_, it is also well suited